lib/storage/azure.js

```diff
@@ -183,54 +183,53 @@ module.exports = function() {
     },
 
     copyIn: function(localPath, _path, options, callback) {
+      const randStr = () => Math.random().toString(36).substring(7);
       const path = _path[0] === '/' ? _path.slice(1) : _path;
-      const tmpFileName = Math.random().toString(36).substring(7);
-
-      // store gzipped file in temp dir
-      fs.mkdtemp(this.options.tempPath, (e, dir) => {
-        if (e) {
-          console.log(e);
-          return callback(e);
-        } else {
-          const tempPath = dir + '/' + tmpFileName;
-          let inp = fs.createReadStream(localPath);
-          let out = fs.createWriteStream(tempPath);
-
-          inp.on('error', e => {
-            console.log("Error in write stream", e);
-            callback(e);
-          });
+      const tmpFileName = randStr();
+      const tempPath = this.options.tempPath + '/' + tmpFileName;
+
+      let inp = fs.createReadStream(localPath);
+      let out = fs.createWriteStream(tempPath);
+      let hasError = false;
 
-          out.on('error', e => {
-            console.log("Error in write stream", e);
-            callback(e);
+      inp.on('error', inpErr => {
+        console.log("Error in read stream", inpErr);
+        hasError = true;
+        callback(inpErr);
+      });
+
+      out.on('error', outErr => {
+        if (outErr && !hasError) {
+          console.log('Error in write stream', outErr);
+          callback(outErr);
+
+          // clean up
+          fs.unlink(tempPath, outErrUnlinkErr => {
```

Comment:
Cleanup should be a function reused in both error handlers. That function could also handle checking the flag to bypass the whole thing, setting the flag, and actually invoking callback.

Reply:
nothing to clean up on a file read

Reply:
I disagree about cleanup method as well since the cleanup cases are different, and adding a helper function interupts the callback flow which is where most of the complexity lies.

Reply:
The cleanup method isn't different though, because we don't really know to which stream node might report an error (or both), and in either case an output file / folder may exist and need removing. That's why I am asking for a single "error cleanup" function which checks for things still being out there and shuts them down / removes them gracefully if they are, if and only if this hasn't already been done.

Reply:
OK, I'll add that. Should make testing easier also, since we can mock the objects and make sure cleanup does the right thing...